invenio_vocabularies/services/service.py

```diff
@@ -37,6 +47,148 @@
 from .tasks import process_datastream
 
 
+def is_custom_vocabulary_type(vocabulary_type, context):
+    """Check if the vocabulary type is a custom vocabulary type."""
+    return vocabulary_type["id"] in current_app.config.get(
+        "VOCABULARIES_CUSTOM_VOCABULARY_TYPES", []
+    )
+
+
+class VocabularyMetadataList(ServiceListResult):
+    """Ensures that vocabulary metadata is returned in the proper format."""
+
+    def __init__(
+        self,
+        service,
+        identity,
+        results,
+        links_tpl=None,
+        links_item_tpl=None,
+    ):
+        """Constructor.
+
+        :params service: a service instance
+        :params identity: an identity that performed the service request
+        :params results: the search results
+        """
+        self._identity = identity
+        self._results = results
+        self._service = service
+        self._links_tpl = links_tpl
+        self._links_item_tpl = links_item_tpl
+
+    def to_dict(self):
+        """Formats result to a dict of hits."""
+        hits = list(self._results)
+
+        for hit in hits:
+            if self._links_item_tpl:
+                hit["links"] = self._links_item_tpl.expand(self._identity, hit)
+
+        res = {
+            "hits": {
+                "hits": hits,
+                "total": len(hits),
+            }
+        }
+
+        if self._links_tpl:
+            res["links"] = self._links_tpl.expand(self._identity, None)
+
+        return res
+
+
+class VocabularyTypeService(Service):
+    """Vocabulary type service."""
+
+    @property
+    def schema(self):
+        """Returns the data schema instance."""
+        return ServiceSchemaWrapper(self, schema=self.config.schema)
+
+    @property
+    def links_item_tpl(self):
+        """Item links template."""
+        return LinksTemplate(
+            self.config.vocabularies_listing_item,
+        )
+
+    @property
+    def custom_vocabulary_names(self):
+        """Checks whether vocabulary is a custom vocabulary."""
+        return current_app.config.get("VOCABULARIES_CUSTOM_VOCABULARY_TYPES", [])
+
+    def search(self, identity):
+        """Search for vocabulary types entries."""
+        self.require_permission(identity, "list_vocabularies")
+
+        vocabulary_types = VocabularyType.query.all()
+
+        # config_vocab_types = current_app.config["INVENIO_VOCABULARY_TYPE_METADATA"]
+        config_vocab_types = current_app.config.get(
+            "INVENIO_VOCABULARY_TYPE_METADATA", {}
+        )
```

Comment:
minor/shelve: we could try to do this "response enhancement" in the marshmallow schema or even the service. Since it might change a lot, I would leave it here and postpone though.